adapters/repos/db/vector/hnsw/insert.go

```diff
@@ -155,18 +155,39 @@ func (h *hnsw) insert(node *vertex, nodeVec []float32) error {
 
 	nodeId := node.id
 
-	// before = time.Now()
-	h.Lock()
-	// m.addBuildingLocking(before)
-	err := h.growIndexToAccomodateNode(node.id, h.logger)
-	if err != nil {
+	// There are 3 different things going on in this block:
+	// 1) A global read lock to check if we need to grow the index and only upgrade to a RW lock if yes
+	// 2) The write to the nodes-slice needs to be behind a Rlock to secure it against growing from another thread. A
+	//    RW-lock is not necessary as the node is not discoverable until the connections with the neighboring nodes have
+	//	  been established.
+	// 3) The sharded-lock secures against searches with an allowList containing the id of the node that is currently
+	//    being added. An object can already be added to the object store before this insert completes and cause a race
+	//    between writing to the nodes-slice and reading all nodes from the allowlist.
+	h.RLock()
+	previousSize := uint64(len(h.nodes))
+	if nodeId >= previousSize {
+		h.RUnlock()
+		h.Lock()
+		if nodeId >= previousSize {
+			err := h.growIndexToAccomodateNode(node.id, h.logger)
+			if err != nil {
+				h.Unlock()
+				return errors.Wrapf(err, "grow HNSW index to accommodate node %d", node.id)
+			}
+		}
+		h.shardedNodeLocks[nodeId%NodeLockStripe].Lock()
+		h.nodes[nodeId] = node
```

Comment:
This line causes a race condition with [this line](https://github.com/weaviate/weaviate/blob/2a54495a8b6cf2a586e06e4e00ac59fcb750f5e7/adapters/repos/db/vector/hnsw/delete.go#L281) when a delete is happening concurrently. It needs to be behind a Lock instead of a RLock.

@abdelr @dirkkul Should we revert this change or is there anything else you'd want me to try?
Upgrading the lock in the `else` doesn't seem like a solution at first glance